libs/pageserver_api/src/shard.rs

```diff
@@ -526,6 +536,135 @@ impl ShardIdentity {
     }
+
+    pub fn key_range_local_size(&self, range: &Range<Key>) -> u32 {
+        let size = self.key_range_local_size_impl(range);
+        if cfg!(debug_assertions) {
+            // The local size of a range should always be <= the literal number of blocks in the range
+            let naive_size = key_range_size(range);
+            if size > naive_size {
+                panic!(
+                    "Calculated local key range size {size} is greater than naive size {naive_size}"
+                )
+            }
+        }
+        size
+    }
+
+    /// Given a range of the keyspace, calculate how many blocks within the range
+    /// are stored on this shard.
+    ///
+    /// If the range spans multiple relations, we can't calculate this (because we don't
+    /// know about the relation in between), so we return u32::MAX -- this acts as a hint
+    /// to callers in compaction that they should start a new layer.
+    pub fn key_range_local_size_impl(&self, range: &Range<Key>) -> u32 {
+        let start = range.start;
+        let end = range.end;
+        // Although only field4 & field6 are included in the hash, if other fields differ then
+        // it would be inaccurate for us to return a block count that assumed they were the same.
+        if end.field1 != start.field1
+            || end.field2 != start.field2
+            || end.field3 != start.field3
+            || end.field4 != start.field4
+        {
+            return u32::MAX;
+        }
+
+        // Fast path: avoid hashing if we can immediately identify the owner of the whole range
+        if self.is_unsharded()
+            || self.count == ShardCount(1)
+            || (key_is_shard0(&start) && self.number == ShardNumber(0))
+        {
+            let start = (start.field5 as u64) << 32 | start.field6 as u64;
+            let end = (end.field5 as u64) << 32 | end.field6 as u64;
+
+            let diff = end - start;
+            if diff > u32::MAX as u64 {
+                return u32::MAX;
+            } else {
+                return diff as u32;
+            }
+        }
+
+        // Special cases for single keys like logical sizes
+        if end == start.add(1) && self.is_key_global(&start) {
+            return 1;
+        }
+
+        // Normal path: step through stripes and part-stripes in the range, evaluate whether each one belongs
+        // to Self, and add the stripe's block count to our total if so.
+        let mut result: u64 = 0;
+        let mut stripe_start = start;
+        while stripe_start < end {
+            let shard = key_to_shard_number(self.count, self.stripe_size, &stripe_start);
+
+            // Count up to the next stripe_size boundary
+            let stripe_index = stripe_start.field6 / self.stripe_size.0;
+            let stripe_remainder =
+                self.stripe_size.0 - (stripe_start.field6 - stripe_index * self.stripe_size.0);
+
+            let next_stripe_start = stripe_start.add(stripe_remainder);
```

Comment:
I think Key::add already has this behavior -- it does an overflowing_add for all the individial fields, and then if there is still an overflow when incrementing field1, it asserts out.

Reply:
Sure. It will assert out on overflow of field1, but we shouldn't overflow after field 5 IIUC (if we do, something's wrong). In any case, this was a nit.

Reply:
ah, I see what you mean.  Added a debug assertion here.